src/pymor/models/iosys.py

```diff
@@ -575,32 +590,39 @@ def gramian(self, typ, mu=None):
 
         assert typ in ('c_lrcf', 'o_lrcf', 'c_dense', 'o_dense')
 
-        if not isinstance(mu, Mu):
-            mu = self.parameters.parse(mu)
-        assert self.parameters.assert_compatible(mu)
-        A = self.A.assemble(mu)
-        B = self.B
-        C = self.C
-        E = self.E.assemble(mu) if not isinstance(self.E, IdentityOperator) else None
-        options_lrcf = self.solver_options.get('lyap_lrcf') if self.solver_options else None
-        options_dense = self.solver_options.get('lyap_dense') if self.solver_options else None
-
-        if typ == 'c_lrcf':
-            return solve_lyap_lrcf(A, E, B.as_range_array(mu=mu),
-                                   trans=False, options=options_lrcf)
-        elif typ == 'o_lrcf':
-            return solve_lyap_lrcf(A, E, C.as_source_array(mu=mu),
-                                   trans=True, options=options_lrcf)
-        elif typ == 'c_dense':
-            return solve_lyap_dense(to_matrix(A, format='dense'),
-                                    to_matrix(E, format='dense') if E else None,
-                                    to_matrix(B, format='dense'),
-                                    trans=False, options=options_dense)
-        elif typ == 'o_dense':
-            return solve_lyap_dense(to_matrix(A, format='dense'),
-                                    to_matrix(E, format='dense') if E else None,
-                                    to_matrix(C, format='dense'),
-                                    trans=True, options=options_dense)
+        if typ in self.presets.keys():
+            return self.presets[typ]
+        elif typ == 'c_lrcf' and 'c_dense' in self.presets.keys():
+            return self.A.source.from_numpy(_chol(self.presets['c_dense']).T)
+        elif typ == 'o_lrcf' and 'o_dense' in self.presets.keys():
+            return self.A.source.from_numpy(_chol(self.presets['o_dense']).T)
+        else:
```

Comment:
Yes this is not nice atm. I would like it, if all the `typ ==` checks could be in one long `if`, `elif` chain. Do you have an idea how to do that?

I can't quite gauge how expensive the `A.assemble` can become and if the `B = self.B` would be handled intelligently by the interpreter. If affordable, these assignments could be moved above the whole `if` clause.

Reply:
I like it as it is now, i.e., first using `presets` as much as possible, and if that fails, doing the usual computation.

If `A` is non-parametric, `A.assemble` doesn't do anything (it should just return `A`). If `A` is parametric, it depends on the structure of `A` what `A.assemble` does. In the simple case where `A` is a `LincombOperator` involving `NumpyMatrixOperators` and parametric coefficients, `A.assemble` would compute the linear combination of `NumpyMatrixOperators`.
As far as I understand, `B = self.B` just adds the name "B" to local variables and points it to the `self.B` object (i.e., `locals()['B'] is self.B`).

Reply:
OK great, then I was worrying too much. I don't see a major drawback in doing the assignments before the `if`,`elif` clause, because if larger computations are involved, the results are always used later. For my taste, the unecessary assignment of local variables is bearable in order to improve readability.